### PR TITLE
update pari to 2.15.4, drop patch

### LINKED ISSUE #35219: Failure to handle large prime input with Pari's `factor()`
### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### Did you read the documentation and troubleshoot guide?

- [X] I have read the documentation and troubleshoot guide

### Environment

```markdown
- **OS**: MacOS
- **Sage Version**: Version 9.8
```

### Steps To Reproduce

When working with large primes, Pari no longer seems to properly handle `factor(large_prime)` and hangs indefinitely. This affects other functionality such as `Mod(2, p).sqrt()` or `Zmod(p)(2).sqrt()`, as they rely on `factor()` under the hood. Note that it does not affect `GF(p)(2).sqrt()`, as Sage already knows the modulus is prime in this case.

For example, defining a large prime and attempting to factor it hangs:

```python
sage: p = 30531851861994333252675935111487950694414332763909083514133769861350960895076504687261369815735742549428789138300843082086550059082835141454526618160634109969195486322015775943030060449557090064811940139431735209185996454739163555910726493597222646855506445602953689527405362207926990442391705014604777038685880527537489845359101552442292804398472642356609304810680731556542002301547846635101455995732584071355903010856718680732337369128498655255277003643669031694516851390505923416710601212618443109844041514942401969629158975457079026906304328749039997262960301209158175920051890620947063936347307238412281568760161
sage: %timeit p.is_prime()
12.4 s ± 477 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
sage: %timeit p.is_prime(proof=False)
8.71 ms ± 166 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
sage: %time factor(p, proof=False)

```

If we set the verbosity to be high we get the following (truncated) output

```python
sage: %time factor(p, proof=False, verbose=8)
IFAC: cracking composite
	30531851861994333252675935111487950694414332763909083514133769861350960895076504687261369815735742549428789138300843082086550059082835141454526618160634109969195486322015775943030060449557090064811940139431735209185996454739163555910726493597222646855506445602953689527405362207926990442391705014604777038685880527537489845359101552442292804398472642356609304810680731556542002301547846635101455995732584071355903010856718680732337369128498655255277003643669031694516851390505923416710601212618443109844041514942401969629158975457079026906304328749039997262960301209158175920051890620947063936347307238412281568760161
IFAC: checking for pure square
OddPwrs: examining 2048-bit integer

OddPwrs: not a power
IFAC: trying Pollard-Brent rho method
Rho: searching small factor of 2048-bit integer
Rho: using X^2+1 for up to 49152 rounds of 32 iterations
Rho: time =   3278 ms,	24576 rounds
Rho: fast forward phase (8192 rounds of 64)...
Rho: time =   1480 ms,	32772 rounds, back to normal mode
Rho: time =    852 ms,	40960 rounds
Rho: time =    872 ms,	Pollard-Brent giving up.
IFAC: trying Shanks' SQUFOF, will fail silently if input
      is too large for it.
IFAC: trying Lenstra-Montgomery ECM
ECM: working on 64 curves at a time; initializing for up to 912 rounds...
ECM: time =      0 ms
ECM: B1 = 1800,	B2 = 198000,	gss =  128*420
ECM: time =   2996 ms, B1 phase done, p = 1801, setting up for B2
	(got [2]Q...[10]Q)
	(got [p]Q, p = 1811 = prc210_rp[29] mod 210)
	(got initial helix)
ECM: time =     55 ms, entering B2 phase, p = 2017
ECM: finishing curves 60...63
#
# TRUNCATED
#
```

**Note**, if we supply an integer which *is* composite with a very large prime factor, Pari seems to work as expected, so the issue may be that there is no primality check at the beginning of the Pari function `factor()`:

```python
sage: %time factor(2 * p, proof=False)
CPU times: user 20.7 ms, sys: 87 µs, total: 20.7 ms
Wall time: 20.7 ms
2 * 30531851861994333252675935111487950694414332763909083514133769861350960895076504687261369815735742549428789138300843082086550059082835141454526618160634109969195486322015775943030060449557090064811940139431735209185996454739163555910726493597222646855506445602953689527405362207926990442391705014604777038685880527537489845359101552442292804398472642356609304810680731556542002301547846635101455995732584071355903010856718680732337369128498655255277003643669031694516851390505923416710601212618443109844041514942401969629158975457079026906304328749039997262960301209158175920051890620947063936347307238412281568760161
```

### Expected Behavior

For prime input to be tested for primality before applying various factoring algorithms

### Actual Behavior

Pari seems to attempt to factor the large integer without checking first that the input is prime and hangs indefinitely as no factors of the prime will be found

### Additional Information

SageMath 9.8 included the ticket #34537, which upgraded PARI to version 2.15.2. Potentially, the factoring error was introduced in this new version.

### PR DESCRIPTION
**Pull request by dimpase**
### 📚 Description

update pari to 2.15.4, drop patch

Will fix #35219 

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.

### CODE REVIEW

**Comment by orlitzky** [comment]
I updated it in ::gentoo. The changelog looks safe but I haven't run the sage test suite yet. The "Build & Test" CI has a green checkmark but looks like it crashed to me?

**Comment by dimpase** [comment]
What has crashed? 
https://github.com/sagemath/sage/actions/runs/4446987505/jobs/7807969635?pr=35302 says "all tests passed"

**Comment by orlitzky** [comment]
It also won't let me upload a screenshot. The raw logs look OK, it's only the summary that's busted.

**Review by orlitzky — APPROVED** [review]

**Comment by yyyyx4** [comment]
On the topic of patching PARI, please have a look at #35219.

File: src/sage/arith/misc.py [hunk 1]
```diff
@@ -2656,6 +2656,11 @@ def factor(n, proof=None, int_=False, algorithm='pari', verbose=0, **kwds):
         Traceback (most recent call last):
         ...
         TypeError: unable to factor 'xyz'
+
+    test that :issue:`35219` is fixed::
```
Comment (yyyyx4) [inline src/sage/arith/misc.py]: "Test" should start with an uppercase letter here.

Comment (yyyyx4) [reply src/sage/arith/misc.py]: (Aside from this single-bit detail, it looks good to me.)

Comment (dimpase) [reply src/sage/arith/misc.py]: ok

**Review by yyyyx4 — APPROVED** [review]

**Comment by vbraun** [comment]
```
File "src/sage/interfaces/gp.py", line 619, in sage.interfaces.gp.Gp._next_var_name
Failed example:
    for n in [1..13000]:  # long time
        a = g(n)          # long time
Exception raised:
    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/interfaces/expect.py", line 523, in _start
        self._expect.expect(self._prompt)
      File "/home/release/Sage/local/var/lib/sage/venv-python3.11.1/lib/python3.11/site-packages/pexpect/spawnbase.py", line 343, in expect
        return self.expect_list(compiled_pattern_list,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/release/Sage/local/var/lib/sage/venv-python3.11.1/lib/python3.11/site-packages/pexpect/spawnbase.py", line 372, in expect_list
        return exp.expect_loop(timeout)
               ^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/release/Sage/local/var/lib/sage/venv-python3.11.1/lib/python3.11/site-packages/pexpect/expect.py", line 179, in expect_loop
        return self.eof(e)
               ^^^^^^^^^^^
      File "/home/release/Sage/local/var/lib/sage/venv-python3.11.1/lib/python3.11/site-packages/pexpect/expect.py", line 122, in eof
        raise exc
    pexpect.exceptions.EOF: End Of File (EOF). Exception style platform.
    PARI/GP interpreter finished running /home/release/Sage/local/bin/gp --fast --emacs --quiet --stacksize 10000
    command: /home/release/Sage/local/bin/gp
    args: ['/home/release/Sage/local/bin/gp', '--fast', '--emacs', '--quiet', '--stacksize', '10000']
    buffer (last 100 chars): b''
    before (last 100 chars): b" [hint] set 'parisizemax' to a nonzero value in your GPRC\r\n\r\n### Errors on startup, exiting...\r\n\r\n\r\n"
    after: <class 'pexpect.exceptions.EOF'>
    match: None
    match_index: None
    exitstatus: 1
    flag_eof: True
    pid: 2983226
    child_fd: 23
    closed: False
    timeout: None
    delimiter: <class 'pexpect.exceptions.EOF'>
    logfile: None
    logfile_read: None
    logfile_send: None
    maxread: 4194304
    ignorecase: False
    searchwindowsize: None
    delaybeforesend: None
    delayafterclose: 0.1
    delayafterterminate: 0.1
    searcher: searcher_re:
        0: re.compile(b'\\? ')
```
And indeed doesn't start with stacksize 10kb:
```
[release@zen Sage]$ /home/release/Sage/local/bin/gp --fast --emacs --quiet --stacksize 10000
  ***   the PARI stack overflows !
  current stack size: 10000 (0.010 Mbytes)
  [hint] set 'parisizemax' to a nonzero value in your GPRC
```

**Comment by dimpase** [comment]
> gp --fast --emacs --quiet --stacksize 10000
works for me, e.g. on macOS:
```
% gp --fast --emacs --stacksize 10000 
                  GP/PARI CALCULATOR Version 2.15.3 (released)
          i386 running darwin (x86-64/GMP-6.2.1 kernel) 64-bit version
    compiled: Mar 14 2023, Apple clang version 14.0.0 (clang-1400.0.29.202)
                           threading engine: pthread
                 (readline v8.2 enabled, extended help enabled)

                     Copyright (C) 2000-2022 The PARI Group

PARI/GP is free software, covered by the GNU General Public License, and comes 
WITHOUT ANY WARRANTY WHATSOEVER.

Type ? for help, \q to quit.
Type ?18 for how to get moral (and possibly technical) support.

parisize = 10000, primelimit = 500000, nbthreads = 8
? 2+2
%1 = 4
? 
```
and on GitHub's CI.
@vbraun - what platform is that `zen` ?

**Comment by yyyyx4** [comment]
Upstream patch is now available at [#2469](https://pari.math.u-bordeaux.fr/cgi-bin/bugreport.cgi?bug=2469). (Is there any way for me to push to this branch?)

**Comment by dimpase** [comment]
everyone can open a PR against the branch of this PR, irrespective of any team membership. I.e. go to my fork and open a PR there, against the  correct base.

**Comment by yyyyx4** [comment]
@orlitzky, would you mind reviewing this again? We should really get the fix for #35219 merged before the next release.

**Comment by orlitzky** [comment]
Better that I don't, I strongly disagree with the practice of backporting every upstream bugfix commit and then testing for it three times.

**Comment by tornaria** [comment]
> Better that I don't, I strongly disagree with the practice of backporting every upstream bugfix commit and then testing for it three times.

I also dislike forcing the user to compile pari instead of using the system version just because of some bug or another. Not like any of these make sagemath unusable, not even close to that.

Also "cannot use system pari/GP with known bug" sounds weird, as in "we could use it before we knew the bug, but now that someone put the cat on the table we have no choice... (blame the messenger)."

 I also agree with @orlitzky that the new doctests are unnecessary since they are not testing code in sagelib. I'm sure pari upstream will add tests.

I'd say just add the patches to the pari spkg and be done with it. Pari/GP is a good upstream, could we just trust them?

**Comment by mkoeppe** [comment]
Any interest in updating this to 2.15.4?

**Comment by dimpase** [comment]
OK, let's see how it goes with .4

**Comment by dimpase** [comment]
I've left tests in (2 new tests in sagelib, and 2 tests in spkg-configure.m4)
These are quick and useful - Pari is unfortunately known for sometimes breaking their own bug fixes...

**Comment by dimpase** [comment]
Please review

**Comment by mkoeppe** [comment]
Tested ([with some other update PRs](https://github.com/mkoeppe/sage/commits/ci-10.2.beta8%2Bupdates-1)) in https://github.com/mkoeppe/sage/actions/runs/6622955701

**Comment by vbraun** [comment]
I'm getting a lot of 
```
**********************************************************************
File "src/sage/rings/integer.pyx", line 1471, in sage.rings.integer.Integer.digits
Failed example:
    n.digits(base=10)[-1]  # slightly slower than str
Exception raised:
    Traceback (most recent call last):
      File "sage/structure/parent.pyx", line 895, in sage.structure.parent.Parent.__call__
        mor = <map.Map> self._convert_from_hash.get(R)
      File "sage/structure/coerce_dict.pyx", line 650, in sage.structure.coerce_dict.MonoDict.get
        raise KeyError(k)
    KeyError: Integer Ring
    During handling of the above exception, another exception occurred:
    Traceback (most recent call last):
      File "sage/structure/parent.pyx", line 2503, in sage.structure.parent.Parent._internal_convert_map_from
        return self._convert_from_hash.get(S)
      File "sage/structure/coerce_dict.pyx", line 650, in sage.structure.coerce_dict.MonoDict.get
        raise KeyError(k)
    KeyError: Integer Ring
    During handling of the above exception, another exception occurred:
    Traceback (most recent call last):
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/doctest/forker.py", line 709, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/doctest/forker.py", line 1144, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.rings.integer.Integer.digits[30]>", line 1, in <module>
        n.digits(base=Integer(10))[-Integer(1)]  # slightly slower than str
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/rings/integer.pyx", line 1523, in sage.rings.integer.Integer.digits
        if do_sig_on: sig_on()
    SystemError: calling remove_from_pari_stack() inside sig_on()
**********************************************************************
```
random, but on half of the buildbots

**Comment by vbraun** [comment]
Also get it without this, probably something else that randomly increases the failure probability.

**Comment by orlitzky** [comment]
> SystemError: calling remove_from_pari_stack() inside sig_on()

- https://github.com/sagemath/sage/issues/28977
- https://github.com/sagemath/sage/issues/30379
- https://github.com/sagemath/sage/issues/33304
- https://github.com/sagemath/sage/issues/33003
- https://github.com/sagemath/sage/issues/33230
- https://github.com/sagemath/sage/issues/28383
- https://github.com/sagemath/sage/issues/33227

**Comment by dimpase** [comment]
I recently removed sign_on/off from a piece of code which was calling plain Python. PR #??? (sorry, afk).
 Perhaps it's more of the same thing.

**Comment by tornaria** [comment]
FWIW, I've been on pari 2.15.3 since march and pari 2.15.4 since july, without any regressions.

Ofc these kind of bugs tend to appear on very heavy loaded boxes. The fact that now appears more often might also be related to the boxes now being more loaded.

**Comment by dimpase** [comment]
this might be related to cypari2 memory model being very suboptimal, creating reference counted Python objects for everything Pari puts on the stack.

**Comment by jhpalmieri** [comment]
OS X does not have a command `timeout`, so that particular test (line 72 in spkg-configure.m4) fails and Sage refuses to use the system `pari`.

**Comment by jhpalmieri** [comment]
Some untested options for OS X: https://unix.stackexchange.com/a/43346, https://unix.stackexchange.com/a/18711, https://stackoverflow.com/a/35512328/1401572.

**Comment by orlitzky** [comment]
> OS X does not have a command `timeout`, so that particular test (line 72 in spkg-configure.m4) fails and Sage refuses to use the system `pari`.

We already have a version test near the top of pari's `spkg-configure.m4`. The additional tests for `factor()` and `qfbclassno()` are there because the main version test allows versions of pari older than 2.15.4. So, if you are extremely lucky/weird and if you have a pari-2.15.3 with two patches that fix bugs 2466 and 2469, then `./configure` would still accept that pari-2.15.3.

I have always thought this was a waste of time, but since in this case it was someone else's time, who am I to complain :)

But what I'm getting at is that maybe now we should just bump the main version requirement up to 2.15.4. Anyone who would've packed such a FrankenPARI-2.15.3 will most certainly have upgraded to 2.15.4 by now. Then the two bug tests (and in particular, the `timeout`) can be deleted.

**Comment by orlitzky** [comment]
```
$ gp --version
                  GP/PARI CALCULATOR Version 2.15.4 (released)
```

```
$ grep -r PARI_VERSION_CODE /usr/include/pari/
/usr/include/pari/paricfg.h:#define PARI_VERSION_CODE 134916
```

**Comment by jhpalmieri** [comment]
> > OS X does not have a command `timeout`, so that particular test (line 72 in spkg-configure.m4) fails and Sage refuses to use the system `pari`.
> 
> We already have a version test near the top of pari's `spkg-configure.m4`. The additional tests for `factor()` and `qfbclassno()` are there because the main version test allows versions of pari older than 2.15.4. So, if you are extremely lucky/weird and if you have a pari-2.15.3 with two patches that fix bugs 2466 and 2469, then `./configure` would still accept that pari-2.15.3.
> 
> I have always thought this was a waste of time, but since in this case it was someone else's time, who am I to complain :)
> 
> But what I'm getting at is that maybe now we should just bump the main version requirement up to 2.15.4. Anyone who would've packaged such a FrankenPARI-2.15.3 will most certainly have upgraded to 2.15.4 by now. Then the two bug tests (and in particular, the `timeout`) can be deleted.

Or the logic could be: if you're using anything older than 2.15.4, run those particular tests, but skip those tests with 2.15.4 or later. Or at least do this with the one test that uses `timeout`.

I don't really have an opinion about how to solve it, I would just like a solution that lets me use homebrew's pari:
```
% brew info pari
==> pari: stable 2.15.4 (bottled)
...
```
and
```
% gp --version
                       GP/PARI CALCULATOR Version 2.15.4 (released)
```

**Comment by orlitzky** [comment]
bumping the lower bound in https://github.com/sagemath/sage/pull/36732

**Pull request merged**